File: app/src/main/java/com/ch/yoon/kakao/pay/imagesearch/ui/imagesearch/imagelist/ImageListViewModel.java
package com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.imagelist;

import android.app.Application;
import android.text.TextUtils;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.ch.yoon.kakao.pay.imagesearch.R;
import com.ch.yoon.kakao.pay.imagesearch.data.model.imagesearch.response.ImageDocument;
import com.ch.yoon.kakao.pay.imagesearch.data.model.imagesearch.response.SearchMetaInfo;
import com.ch.yoon.kakao.pay.imagesearch.data.repository.ImageRepository;
import com.ch.yoon.kakao.pay.imagesearch.data.model.imagesearch.request.ImageSearchRequest;
import com.ch.yoon.kakao.pay.imagesearch.data.model.imagesearch.request.ImageSortType;
import com.ch.yoon.kakao.pay.imagesearch.data.model.imagesearch.response.error.ImageSearchError;
import com.ch.yoon.kakao.pay.imagesearch.data.model.imagesearch.response.error.ImageSearchException;
import com.ch.yoon.kakao.pay.imagesearch.ui.base.BaseViewModel2;
import com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.imagelist.helper.ImageSearchInspector;
import com.ch.yoon.kakao.pay.imagesearch.utils.CollectionUtil;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;

/**
 * Creator : ch-yoon
 * Date : 2019-08-01.
 */
public class ImageListViewModel extends BaseViewModel2 {

    private static final String TAG = ImageListViewModel.class.getName();
    private static final ImageSortType DEFAULT_IMAGE_SORT_TYPE = ImageSortType.ACCURACY;
    private static final int DEFAULT_COUNT_OF_ITEM_IN_LINE = 2;

    @NonNull
    private final ImageRepository imageRepository;
    @NonNull
    private final ImageSearchInspector imageSearchInspector;

    @NonNull
    private final MutableLiveData<ImageSortType> imageSortTypeLiveData = new MutableLiveData<>();
    @NonNull
    private final MutableLiveData<Integer> countOfItemInLineLiveData = new MutableLiveData<>();
    @NonNull
    private final MutableLiveData<List<ImageDocument>> imageDocumentListLiveData = new MutableLiveData<>();
    @NonNull
    private final MutableLiveData<ImageSearchState> imageSearchStateLiveData = new MutableLiveData<>();

    @Nullable
    private SearchMetaInfo searchMetaInfo;

    public ImageListViewModel(@NonNull Application application,
                       @NonNull ImageRepository imageRepository,
                       @NonNull ImageSearchInspector imageSearchInspector) {
        super(application);
        this.imageRepository = imageRepository;
        this.imageSearchInspector = imageSearchInspector;

        init();
    }

    private void init() {
        imageSortTypeLiveData.setValue(DEFAULT_IMAGE_SORT_TYPE);
        countOfItemInLineLiveData.setValue(DEFAULT_COUNT_OF_ITEM_IN_LINE);
        imageSearchStateLiveData.setValue(ImageSearchState.NONE);
        observeImageSearchApprove();
    }

    @NonNull
    public LiveData<ImageSortType> observeImageSortType() {
        return imageSortTypeLiveData;
    }

    @NonNull
    public LiveData<Integer> observeCountOfItemInLine() {
        return countOfItemInLineLiveData;
    }

    @NonNull
    public LiveData<List<ImageDocument>> observeImageDocumentList() {
        return imageDocumentListLiveData;
    }

    @NonNull
    public LiveData<ImageSearchState> observeImageSearchState() {
        return imageSearchStateLiveData;
    }

    public void changeImageSortType(final ImageSortType imageSortType) {
        clearBeforeImageDocumentList();
        imageSortTypeLiveData.setValue(imageSortType);

        final String previousKeyword = imageSearchInspector.getPreviousRequestKeyword();
        if(!TextUtils.isEmpty(previousKeyword)) {
            imageSearchInspector.submitFirstImageSearchRequest(previousKeyword, getImageSortType());
        }
    }

    public void changeCountOfItemInLine(final int countOfItemInLine) {
        countOfItemInLineLiveData.setValue(countOfItemInLine);
    }

    public void loadImageList(@NonNull final String keyword) {
        clearBeforeImageDocumentList();
        imageSearchInspector.submitFirstImageSearchRequest(keyword, getImageSortType());
    }

    public void retryLoadMoreImageList() {
        imageSearchInspector.submitRetryRequest(DEFAULT_IMAGE_SORT_TYPE);
    }

    public void loadMoreImageListIfPossible(final int displayPosition) {
        if(isRemainingMoreData()) {
            final List<ImageDocument> imageDocumentList = imageDocumentListLiveData.getValue();
            imageSearchInspector.submitPreloadRequest(
                displayPosition,
                imageDocumentList != null ? imageDocumentList.size() : 0,
                getImageSortType(),
                getCountOfItemInLine()
            );
        }
    }

    private void observeImageSearchApprove() {
        imageSearchInspector.observeImageSearchApprove(this::requestImageSearchToRepository);
    }

    private void requestImageSearchToRepository(final ImageSearchRequest imageSearchRequest) {
        changeImageSearchState(ImageSearchState.NONE);

        registerDisposable(
            imageRepository.requestImageList(imageSearchRequest)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(imageSearchResult -> {
                    changeImageSearchState(ImageSearchState.SUCCESS);
                    clearBeforeDocumentListIfFirstRequest(imageSearchResult.getImageSearchRequest());
                    updateSearchMetaInfo(imageSearchResult.getSearchMetaInfo());
                    updateImageInfoList(imageSearchResult.getImageDocumentList());
                }, throwable -> {
                    changeImageSearchState(ImageSearchState.FAIL);
                    if(throwable instanceof ImageSearchException) {
                        final ImageSearchError error = ((ImageSearchException)throwable).getImageSearchError();
                        updateMessage(error.getErrorMessageResourceId());
                    } else {
                        Log.d(TAG, throwable.getMessage());
                    }
                })
        );
    }

    private void clearBeforeDocumentListIfFirstRequest(final ImageSearchRequest imageSearchRequest) {
        if(imageSearchRequest.isFirstRequest()) {
            clearBeforeImageDocumentList();
        }
    }

    private void updateSearchMetaInfo(final SearchMetaInfo searchMetaInfo) {
        this.searchMetaInfo = searchMetaInfo;
    }

    private void updateImageInfoList(final List<ImageDocument> receivedImageDocumentList) {
        final List<ImageDocument> oldList = imageDocumentListLiveData.getValue();
        final List<ImageDocument> newList = new ArrayList<>(oldList == null ? new ArrayList<>() : oldList);
        newList.addAll(receivedImageDocumentList);

        if(CollectionUtil.isEmpty(newList)) {
            updateMessage(R.string.success_image_search_no_result);
        } else if(isNotRemainingMoreData()){
            updateMessage(R.string.success_image_search_last_data);
        }

        imageDocumentListLiveData.setValue(newList);
    }

    private void clearBeforeImageDocumentList() {
        imageDocumentListLiveData.setValue(null);
    }

    private boolean isNotRemainingMoreData() {
        return !isRemainingMoreData();
    }

    private boolean isRemainingMoreData() {
        return searchMetaInfo == null || !searchMetaInfo.isEnd();
    }

    private void changeImageSearchState(final ImageSearchState imageSearchState) {
        final ImageSearchState previousImageSearchState = imageSearchStateLiveData.getValue();
        if(previousImageSearchState != imageSearchState) {
            imageSearchStateLiveData.setValue(imageSearchState);
        }
    }

    private ImageSortType getImageSortType() {
        final ImageSortType imageSortType = imageSortTypeLiveData.getValue();
        return imageSortType != null ? imageSortType : DEFAULT_IMAGE_SORT_TYPE;
    }

    private int getCountOfItemInLine() {
        final Integer countOfItemInLine = countOfItemInLineLiveData.getValue();
        return countOfItemInLine != null ? countOfItemInLine : DEFAULT_COUNT_OF_ITEM_IN_LINE;
    }

}

File: app/src/main/java/com/ch/yoon/kakao/pay/imagesearch/ui/imagesearch/imagelist/ImageListViewModel.kt
package com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.imagelist

import android.app.Application
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import com.ch.yoon.kakao.pay.imagesearch.R
import com.ch.yoon.kakao.pay.imagesearch.data.repository.error.RepositoryException
import com.ch.yoon.kakao.pay.imagesearch.data.source.remote.kakao.request.ImageSearchRequest
import com.ch.yoon.kakao.pay.imagesearch.data.source.remote.kakao.request.ImageSortType
import com.ch.yoon.kakao.pay.imagesearch.data.repository.image.ImageRepository
import com.ch.yoon.kakao.pay.imagesearch.data.repository.image.model.ImageDocument
import com.ch.yoon.kakao.pay.imagesearch.data.repository.image.model.ImageSearchMeta
import com.ch.yoon.kakao.pay.imagesearch.data.repository.image.model.ImageSearchResponse
import com.ch.yoon.kakao.pay.imagesearch.extention.TAG
import com.ch.yoon.kakao.pay.imagesearch.extention.safeLet
import com.ch.yoon.kakao.pay.imagesearch.extention.updateOnMainThread
import com.ch.yoon.kakao.pay.imagesearch.ui.base.BaseViewModel
import com.ch.yoon.kakao.pay.imagesearch.ui.common.livedata.SingleLiveEvent
import com.ch.yoon.kakao.pay.imagesearch.ui.common.pageload.PageLoadHelper
import io.reactivex.android.schedulers.AndroidSchedulers
import java.util.ArrayList

/**
 * Creator : ch-yoon
 * Date : 2019-10-28
 **/
class ImageListViewModel(
    application: Application,
    private val imageRepository: ImageRepository,
    private val pageLoadHelper: PageLoadHelper<String>
) : BaseViewModel(application) {

    init {
        observePageLoadInspector()
    }

    private val _imageSortType = MutableLiveData(ImageSortType.ACCURACY)
    val imageSortType: LiveData<ImageSortType> = _imageSortType

    private val _countOfItemInLine = MutableLiveData(2)
    val countOfItemInLine: LiveData<Int> = _countOfItemInLine

    private val _imageDocumentList = MutableLiveData<MutableList<ImageDocument>>()
    val kakaoImageDocumentList: LiveData<List<ImageDocument>> = Transformations.map(_imageDocumentList) { it?.toList() }

    private val _imageSearchState = MutableLiveData<ImageSearchState>(ImageSearchState.NONE)
    val imageSearchState: LiveData<ImageSearchState> = _imageSearchState

    private val _moveToDetailScreenEvent = SingleLiveEvent<ImageDocument>()
    val moveToDetailScreenEvent: LiveData<ImageDocument> = _moveToDetailScreenEvent

    private val isNotRemainingMoreData
        get() = isRemainingMoreData.not()

    private val isRemainingMoreData
        get() = searchMeta?.isEnd?.not() ?: true

    private var searchMeta: ImageSearchMeta? = null

    fun changeImageSortType(imageSortType: ImageSortType) {
        _imageDocumentList.value = null
        _imageSortType.value = imageSortType
        pageLoadHelper.requestStartOverFromTheBeginning()
    }

    fun changeCountOfItemInLine(countOfItemInLine: Int) {
        _countOfItemInLine.value = countOfItemInLine
    }

    fun loadImageList(keyword: String) {
        _imageDocumentList.value = null
        pageLoadHelper.requestFirstLoad(keyword)
    }

    fun retryLoadMoreImageList() {
        pageLoadHelper.requestRetryAsPreviousValue()
    }

    fun loadMoreImageListIfPossible(displayPosition: Int) {
        if (isRemainingMoreData) {
            safeLet(_countOfItemInLine.value, _imageDocumentList.value) { count, documents ->
                pageLoadHelper.requestPreloadIfPossible(displayPosition, documents.size, count)
            }
        }
    }

    fun onClickImage(imageDocument: ImageDocument) {
        _moveToDetailScreenEvent.value = imageDocument
    }

    private fun observePageLoadInspector() {
        pageLoadHelper.onPageLoadApprove = { key, pageNumber, dataSize, isFirstPage ->
            _imageSortType.value?.let { sortType ->
                val request = ImageSearchRequest(key, sortType, pageNumber, dataSize, isFirstPage)
                requestImageSearchToRepository(request)
            }
        }
    }

    private fun requestImageSearchToRepository(imageSearchRequest: ImageSearchRequest) {
        _imageSearchState.value = ImageSearchState.NONE

        imageRepository.requestImageList(imageSearchRequest)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ imageSearchResponse ->
                _imageSearchState.value = ImageSearchState.SUCCESS
                clearBeforeDocumentListIfFirstRequest(imageSearchRequest)
                handlingImageSearchResult(imageSearchResponse)
            }, { throwable ->
                _imageSearchState.value = ImageSearchState.FAIL
                handlingImageSearchError(throwable)
            })
            .register()
    }

    private fun handlingImageSearchResult(imageSearchResponse: ImageSearchResponse) {
        with(imageSearchResponse) {
            updateSearchMetaInfo(imageSearchMeta)
            updateImageDocumentList(imageDocumentList)
        }
    }

    private fun clearBeforeDocumentListIfFirstRequest(imageSearchRequest: ImageSearchRequest) {
        if (imageSearchRequest.isFirstRequest) {
            _imageDocumentList.value = null
        }
    }

    private fun updateSearchMetaInfo(kakaoImageSearchMetaInfo: ImageSearchMeta) {
        searchMeta = kakaoImageSearchMetaInfo
    }

    private fun updateImageDocumentList(receivedImageDocumentList: List<ImageDocument>) {
        _imageDocumentList.updateOnMainThread { oldList ->
            val newList = oldList ?: ArrayList()
            newList.addAll(receivedImageDocumentList)
            if(newList.isEmpty()) {
                updateShowMessage(R.string.success_image_search_no_result)
            } else if(isNotRemainingMoreData) {
                updateShowMessage(R.string.success_image_search_last_data)
            }
            newList
        }
    }

    private fun handlingImageSearchError(throwable: Throwable) {
        if(throwable is RepositoryException.NetworkNotConnectingException) {
            updateShowMessage(R.string.network_not_connecting_error)
        } else {
            updateShowMessage(R.string.unknown_error)
            Log.d(TAG, throwable.message ?: "unknown error")
        }
    }

}